### PR TITLE
Add enableFeedback param to MaterialButton, RawMaterialButton and IconButton

### PR DESCRIPTION
**Pull request by NikitaZhelonkin**
In some cases developers need to disable default tap feedback, e.g. if you want to implement custom feedback. This PR gives that opportunity

### CODE REVIEW

**Comment by NikitaZhelonkin** [comment]
@googlebot I signed it!

**Comment by shihaohong** [comment]
Hi, this change looks good! Could you add a few tests to verify that the `enableFeedback` parameter is working properly for `MaterialButton` and `RawMaterialButton`? There should be some useful pointers in `ink_well_test.dart`. 

Thank you!

**Comment by NikitaZhelonkin** [comment]
How can i fix " Flutter build is currently broken" ?

**Comment by shihaohong** [comment]
That isn't an issue with your PR, that basically checks if Flutter's latest master is failing for whatever reason. This prevents new changes from being introduced if there is a potential existing issue in the codebase.

File: packages/flutter/lib/src/material/icon_button.dart [hunk 1]
```diff
@@ -269,6 +270,16 @@ class IconButton extends StatelessWidget {
   /// used for accessibility.
   final String tooltip;
 
+  /// Whether detected gestures should provide acoustic and/or haptic feedback.
+  ///
+  /// For example, on Android a tap will produce a clicking sound and a
+  /// long-press will produce a short vibration, when feedback is enabled.
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/icon_button.dart]: It seems that `RawMaterialButton`, `MaterialButton` and `IconButton` do not have an `onLongPress` callback or option, which makes this inaccurate. We should either modify the docs to explain that long-press is not supported for these widgets, or implement them correctly. 

@HansMuller What seems like the appropriate step here?

Comment (shihaohong) [reply packages/flutter/lib/src/material/icon_button.dart]: Related to this issue somewhat, since it explains the default long press behavior for Material buttons and the intent to add an `onLongPress` param: https://github.com/flutter/flutter/pull/40641

File: packages/flutter/lib/src/material/material_button.dart [hunk 2]
```diff
@@ -341,13 +342,24 @@ class MaterialButton extends StatelessWidget {
   /// Defaults to the value from the current [ButtonTheme].
   final double height;
 
+  /// Whether detected gestures should provide acoustic and/or haptic feedback.
+  ///
+  /// For example, on Android a tap will produce a clicking sound and a
+  /// long-press will produce a short vibration, when feedback is enabled.
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/material_button.dart]: Same comment regarding long-press

File: packages/flutter/test/material/buttons_test.dart [hunk 3]
```diff
@@ -1191,4 +1192,45 @@ void main() {
     final Material material = tester.widget<Material>(rawButtonMaterial);
     expect(material.color, const Color(0xff00ff00));
   });
+
+  group('feedback', () {
```
Comment (shihaohong) [inline packages/flutter/test/material/buttons_test.dart]: Could you add a set of these tests for `RawMaterialButton` as well?

File: packages/flutter/test/material/buttons_test.dart [hunk 4]
```diff
@@ -1191,4 +1192,45 @@ void main() {
...
+          onPressed: () {},
+        ),
+      ));
+      await tester.tap(find.byType(MaterialButton), pointer: 1);
+      await tester.pump(const Duration(seconds: 1));
+      expect(feedback.clickSoundCount, 1);
+      expect(feedback.hapticCount, 0);
+    });
+  });
+
```
Comment (shihaohong) [inline packages/flutter/test/material/buttons_test.dart]: nit: remove these two additional newlines

File: packages/flutter/test/material/icon_button_test.dart [hunk 5]
```diff
@@ -397,6 +398,56 @@ void main() {
     expect(focusNode1.hasPrimaryFocus, isTrue);
     expect(focusNode2.hasPrimaryFocus, isFalse);
   });
+
```
Comment (shihaohong) [inline packages/flutter/test/material/icon_button_test.dart]: nit: remove one newline, since two were added

File: packages/flutter/test/material/buttons_test.dart [hunk 6]
```diff
@@ -1191,4 +1192,45 @@ void main() {
...
+          enableFeedback: false,
+        ),
+      ));
+      await tester.tap(find.byType(MaterialButton), pointer: 1);
+      await tester.pump(const Duration(seconds: 1));
+      expect(feedback.clickSoundCount, 0);
+      expect(feedback.hapticCount, 0);
+    });
+
+    testWidgets('MaterialButton with enabled feedback', (WidgetTester tester) async {
```
Comment (shihaohong) [inline packages/flutter/test/material/buttons_test.dart]: This feels like it tests the default behavior (that `enableFeedback` is set to `true` by default). You should probably add a test where `enableFeedback` is explicitly set to `true`. ie.
```
      await tester.pumpWidget(Directionality(
        textDirection: TextDirection.ltr,
        child: MaterialButton(
          onPressed: () {},
          enableFeedback: true, // add a test for this case
        ),
      ));
```

File: packages/flutter/test/material/icon_button_test.dart [hunk 7]
```diff
@@ -397,6 +398,56 @@ void main() {
...
+          ),
+        ),
+      ));
+      await tester.tap(find.byType(IconButton), pointer: 1);
+      await tester.pump(const Duration(seconds: 1));
+      expect(feedback.clickSoundCount, 0);
+      expect(feedback.hapticCount, 0);
+    });
+
+    testWidgets('IconButton with enabled feedback', (WidgetTester tester) async {
```
Comment (shihaohong) [inline packages/flutter/test/material/icon_button_test.dart]: Same comment about testing for explicitly setting `enableFeedback: true`.

**Review by shihaohong — COMMENTED** [review]
Looks pretty good! I have a request about adding an additional test case, and we should wait to see what Hans has to say about handling long presses with material buttons.

File: packages/flutter/lib/src/material/icon_button.dart (hunk 1 above)
Comment (NikitaZhelonkin) [reply packages/flutter/lib/src/material/icon_button.dart]: Removed comment in doc about long-press

File: packages/flutter/lib/src/material/material_button.dart (hunk 2 above)
Comment (NikitaZhelonkin) [reply packages/flutter/lib/src/material/material_button.dart]: Removed comment in doc about long-press

File: packages/flutter/test/material/buttons_test.dart (hunk 3 above)
Comment (NikitaZhelonkin) [reply packages/flutter/test/material/buttons_test.dart]: Done

File: packages/flutter/test/material/buttons_test.dart (hunk 4 above)
Comment (NikitaZhelonkin) [reply packages/flutter/test/material/buttons_test.dart]: Done

File: packages/flutter/test/material/icon_button_test.dart (hunk 5 above)
Comment (NikitaZhelonkin) [reply packages/flutter/test/material/icon_button_test.dart]: Done

File: packages/flutter/test/material/buttons_test.dart (hunk 6 above)
Comment (NikitaZhelonkin) [reply packages/flutter/test/material/buttons_test.dart]: Done

File: packages/flutter/test/material/icon_button_test.dart (hunk 7 above)
Comment (NikitaZhelonkin) [reply packages/flutter/test/material/icon_button_test.dart]: Done

File: packages/flutter/lib/src/material/icon_button.dart (hunk 1 above)
Comment (shihaohong) [reply packages/flutter/lib/src/material/icon_button.dart]: I don't think just removing it is the right approach -- I think if we take the route of not having it respond to long presses, we need to explain how this is different from what is expected. Let's wait for @HansMuller before we continue forward about this topic

**Review by shihaohong — COMMENTED** [review]
The rest of the changes look good, let's just wait to hear back about the long press detail :)

File: packages/flutter/lib/src/material/icon_button.dart (hunk 1 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/icon_button.dart]: I think that #40641 will land pretty soon and so it's OK to explain that enableFeedback does what InkWell.enableFeedback does without qualification.

Comment (NikitaZhelonkin) [reply packages/flutter/lib/src/material/icon_button.dart]: Done

File: packages/flutter/lib/src/material/icon_button.dart [hunk 8]
```diff
@@ -314,15 +325,17 @@ class IconButton extends StatelessWidget {
...
-        focusColor: focusColor ?? Theme.of(context).focusColor,
-        hoverColor: hoverColor ?? Theme.of(context).hoverColor,
-        highlightColor: highlightColor ?? Theme.of(context).highlightColor,
-        splashColor: splashColor ?? Theme.of(context).splashColor,
-        radius: math.max(
-          Material.defaultSplashRadius,
-          (iconSize + math.min(padding.horizontal, padding.vertical)) * 0.7,
-          // x 0.5 for diameter -> radius and + 40% overflow derived from other Material apps.
+        enableFeedback: enableFeedback,
+          child: result,
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/icon_button.dart]: One of the merges with `master` cluttered up this section of code. It should be the original, adding the enableFeedback param.

File: packages/flutter/lib/src/material/material_button.dart (hunk 2 above)
Comment (shihaohong) [reply packages/flutter/lib/src/material/material_button.dart]: Commenting here that we've re-added the comment since long-press is now enabled on these buttons

File: packages/flutter/test/material/buttons_test.dart [hunk 9]
```diff
@@ -616,141 +616,4 @@ void main() {
...
-        ),
-      ),
-    );
-
-    final Material material = tester.widget<Material>(rawButtonMaterial);
-    expect(material.color, const Color(0xff00ff00));
-  });
-
-  group('feedback', () {
-    FeedbackTester feedback;
```
Comment (NikitaZhelonkin) [inline packages/flutter/test/material/buttons_test.dart]: @shihaohong  Why did you remove those tests?

Comment (shihaohong) [reply packages/flutter/test/material/buttons_test.dart]: It was causing a merge conflict since it's part of an ongoing effort to remove the `buttons_test.dart` file. There are more specific files that some of these tests belong to (ie. `flat_button_test.dart`), and it made more sense to have them there.

However, if I seem to have removed a test accidentally that wasn't intentional, please let me know.

**Comment by shihaohong** [comment]
Hi, any updates on this PR? I just want to make sure we can get this one moving along since it's so close to the finish line!

**Comment by shihaohong** [comment]
Hi, any updates on this PR? I just want to make sure we can get this one moving along since it's so close to the finish line!

**Comment by NikitaZhelonkin** [comment]
Fixed formatting; Seems some checks don't passed after merge master of origin repository

**Comment by shihaohong** [comment]
It seems to be failing the following test:
```
   info • Unused import: 'feedback_tester.dart' • packages/flutter/test/material/buttons_test.dart:13:8 • unused_import

```

File: packages/flutter/test/material/buttons_test.dart [hunk 10]
```diff
@@ -10,6 +10,7 @@ import 'package:flutter_test/flutter_test.dart';
 
 import '../rendering/mock_canvas.dart';
 import '../widgets/semantics_tester.dart';
+import 'feedback_tester.dart';
```
Comment (shihaohong) [inline packages/flutter/test/material/buttons_test.dart]: Remove this line

**Comment by shihaohong** [comment]
Make sure to run `flutter analyze --flutter-repo` locally after your changes to make sure that all analyzer issues are properly addressed. 

```
▌17:09:49▐ RUNNING: cd .; grep --line-number --extended-regexp [[:blank:]]$ packages/flutter/lib/src/material/button.dart packages/flutter/lib/src/material/icon_button.dart packages/flutter/lib/src/material/material_button.dart packages/flutter/test/material/icon_button_test.dart
packages/flutter/lib/src/material/material_button.dart:377:      enableFeedback: enableFeedback,      
▌17:09:49▐ ELAPSED TIME: 0.025s for grep --line-number --extended-regexp [[:blank:]]$ packages/flutter/lib/src/material/button.dart packages/flutter/lib/src/material/icon_button.dart packages/flutter/lib/src/material/material_button.dart packages/flutter/test/material/icon_button_test.dart in .
Detected trailing whitespace in the file[s] listed above.
```

**Comment by NikitaZhelonkin** [comment]
It shows '36816 issues found." if i run it locally

File: packages/flutter/lib/src/material/material_button.dart [hunk 11]
```diff
@@ -363,6 +374,7 @@ class MaterialButton extends StatelessWidget {
     return RawMaterialButton(
       onPressed: onPressed,
       onLongPress: onLongPress,
+      enableFeedback: enableFeedback,
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/material_button.dart]:
```suggestion
      enableFeedback: enableFeedback,
```

**Comment by shihaohong** [comment]
Is there any more detail as to what the issues are? Make sure that you've run `flutter update-packages`. It might be that your local copy of Flutter's dependencies are not available.

**Comment by NikitaZhelonkin** [comment]
Yes, `flutter update-packages` helped, thank you!

**Review by shihaohong — APPROVED** [review]
LGTM
